### PR TITLE
Create model and controller fixtures

### PR DESCRIPTION
**Pull request by samokiss**

### CODE REVIEW

File: tests/Fixtures/Classes/Api/Controller/GetFunctionalEntitiesController.php [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+use OpenClassrooms\CodeGenerator\Tests\Fixtures\Classes\BusinessRules\Responders\PaginatedUseCaseResponse;
+use OpenClassrooms\CodeGenerator\Tests\Fixtures\Classes\BusinessRules\UseCases\Domain\SubDomain\GetFunctionalEntities;
+use Symfony\Component\HttpFoundation\JsonResponse;
+
+class GetFunctionalEntitiesController extends AbstractApiController
+{
+    /**
+     * @Security("")
+     */
+    public function getAction(int $userId): JsonResponse
```
Comment (arnaud-23) [inline tests/Fixtures/Classes/Api/Controller/GetFunctionalEntitiesController.php]:
```suggestion
    public function getAction(CollectionInformation $collectionInformation, int $userId): JsonResponse
```

File: tests/Fixtures/Classes/Api/Controller/GetFunctionalEntitiesController.php [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+use OpenClassrooms\CodeGenerator\Tests\Fixtures\Classes\BusinessRules\Requestors\Domain\SubDomain\GetFunctionalEntitiesRequestBuilder;
+use OpenClassrooms\CodeGenerator\Tests\Fixtures\Classes\BusinessRules\Responders\Domain\SubDomain\FunctionalEntityResponse;
+use OpenClassrooms\CodeGenerator\Tests\Fixtures\Classes\BusinessRules\Responders\PaginatedUseCaseResponse;
+use OpenClassrooms\CodeGenerator\Tests\Fixtures\Classes\BusinessRules\UseCases\Domain\SubDomain\GetFunctionalEntities;
+use Symfony\Component\HttpFoundation\JsonResponse;
+
+class GetFunctionalEntitiesController extends AbstractApiController
+{
+    /**
+     * @Security("")
```
Comment (arnaud-23) [inline tests/Fixtures/Classes/Api/Controller/GetFunctionalEntitiesController.php]:
```suggestion
     * @Security("")
     * @ParamConverter("collectionInformation", options={"itemsPerPage":"200"}) // default 20 to remove if not necessary

```

File: tests/Fixtures/Classes/Api/Controller/GetFunctionalEntitiesController.php [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+     */
+    public function getAction(int $userId): JsonResponse
+    {
+        try {
+            $functionalEntities = $this->getFunctionalEntities();
```
Comment (arnaud-23) [inline tests/Fixtures/Classes/Api/Controller/GetFunctionalEntitiesController.php]:
```suggestion
            $functionalEntities = $this->getFunctionalEntities($collectionInformation);
```

File: tests/Fixtures/Classes/Api/Controller/GetFunctionalEntitiesController.php [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+            return $this->createJsonResponse($vm);
+        } catch (FunctionalEntityNotFoundException $e) {
+            throw $this->createNotFoundException();
+        }
+    }
+
+    /**
+     * @throws \OpenClassrooms\CodeGenerator\Tests\Fixtures\Classes\BusinessRules\Gateways\Domain\SubDomain\Exceptions\FunctionalEntityNotFoundException
+     */
+    private function getFunctionalEntities(): FunctionalEntityResponse
```
Comment (arnaud-23) [inline tests/Fixtures/Classes/Api/Controller/GetFunctionalEntitiesController.php]:
```suggestion
    private function getFunctionalEntities(CollectionInformation $collectionInformation): FunctionalEntityResponse
```

File: tests/Fixtures/Classes/Api/Controller/GetFunctionalEntitiesController.php [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    {
+        return $this->get(GetFunctionalEntities::class)->execute(
+            $this->get(GetFunctionalEntitiesRequestBuilder::class)
+                ->create()
+                ->withFilters()
+                ->withItemsPerPage()
+                ->withPage()
+                ->withSort()
```
Comment (arnaud-23) [inline tests/Fixtures/Classes/Api/Controller/GetFunctionalEntitiesController.php]:
```
                ->withFilters($collectionInformation->getFilters())
                ->withItemsPerPage($collectionInformation->getItemsPerPage())
                ->withPage($collectionInformation->getPage())
                ->withSort($collectionInformation->getSorts())
```

File: tests/Fixtures/Classes/Api/Controller/GetFunctionalEntitiesController.php [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+                ->build()
+        );
+    }
+
+    /**
+     * @return FunctionalEntityViewModelListItem[]
+     */
+    private function buildViewModel(PaginatedUseCaseResponse $functionalEntities): array
+    {
+        return $this->get(FunctionalEntityViewModelListItemAssembler::class)->createListItems(
```
Comment (arnaud-23) [inline tests/Fixtures/Classes/Api/Controller/GetFunctionalEntitiesController.php]: this class should be passed from the constructor like the builder on the delete controller here
https://github.com/OpenClassrooms/CodeGenerator/pull/29/files#diff-cc28ade8938bd3062b8e8cf550265a79R21

File: tests/Fixtures/Classes/Api/Controller/GetFunctionalEntitiesController.php [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+        }
+    }
+
+    /**
+     * @throws \OpenClassrooms\CodeGenerator\Tests\Fixtures\Classes\BusinessRules\Gateways\Domain\SubDomain\Exceptions\FunctionalEntityNotFoundException
+     */
+    private function getFunctionalEntities(): FunctionalEntityResponse
+    {
+        return $this->get(GetFunctionalEntities::class)->execute(
+            $this->get(GetFunctionalEntitiesRequestBuilder::class)
```
Comment (arnaud-23) [inline tests/Fixtures/Classes/Api/Controller/GetFunctionalEntitiesController.php]: this class should be passed from the constructor like the builder on the delete controller here
https://github.com/OpenClassrooms/CodeGenerator/pull/29/files#diff-cc28ade8938bd3062b8e8cf550265a79R21

File: tests/Fixtures/Classes/Api/Controller/GetFunctionalEntityController.php [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+        }
+    }
+
+    /**
+     * @throws \OpenClassrooms\CodeGenerator\Tests\Fixtures\Classes\BusinessRules\Gateways\Domain\SubDomain\Exceptions\FunctionalEntityNotFoundException
+     */
+    private function getFunctionalEntity(int $functionalEntityId): FunctionalEntityResponse
+    {
+        return $this->get(GetFunctionalEntity::class)->execute(
+            $this->get(GetFunctionalEntityRequestBuilder::class)
```
Comment (arnaud-23) [inline tests/Fixtures/Classes/Api/Controller/GetFunctionalEntityController.php]: this class should be passed from the constructor like the builder on the delete controller here
https://github.com/OpenClassrooms/CodeGenerator/pull/29/files#diff-cc28ade8938bd3062b8e8cf550265a79R21

File: tests/Fixtures/Classes/Api/Controller/GetFunctionalEntityController.php [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+                ->create()
+                ->withFunctionalEntityId($functionalEntityId)
+                ->build()
+        );
+    }
+
+    protected function createViewModel(FunctionalEntityResponse $functionalEntity): FunctionalEntityViewModel
+    {
+        return $this->get(FunctionalEntityDetailResponseAssembler::class)->create($functionalEntity);
```
Comment (arnaud-23) [inline tests/Fixtures/Classes/Api/Controller/GetFunctionalEntityController.php]: this class should be passed from the constructor like the builder on the delete controller here
https://github.com/OpenClassrooms/CodeGenerator/pull/29/files#diff-cc28ade8938bd3062b8e8cf550265a79R21

File: tests/Fixtures/Classes/Api/Controller/GetFunctionalEntityController.php [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+            $vm = $this->createViewModel($functionalEntity);
+
+            return $this->createJsonResponse($vm);
+        } catch (FunctionalEntityNotFoundException $e) {
+            throw $this->createNotFoundException();
+        }
+    }
+
+    /**
+     * @throws \OpenClassrooms\CodeGenerator\Tests\Fixtures\Classes\BusinessRules\Gateways\Domain\SubDomain\Exceptions\FunctionalEntityNotFoundException
```
Comment (arnaud-23) [inline tests/Fixtures/Classes/Api/Controller/GetFunctionalEntityController.php]: put this long stuff in the "use ...;" part

File: tests/Fixtures/Classes/Api/Controller/PatchFunctionalEntityController.php [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+{
+    /**
+     * @Security("")
+     */
+    public function patchAction(Request $request, int $functionalEntityId): JsonResponse
+    {
+        try {
+            $this->checkContentIsNotEmpty($request);
+
+            $model = new PatchFunctionalEntity($request->getContent());
```
Comment (arnaud-23) [inline tests/Fixtures/Classes/Api/Controller/PatchFunctionalEntityController.php]: PatchFunctionalEntityModel

File: tests/Fixtures/Classes/Api/Controller/PatchFunctionalEntityController.php [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+            $this->validate($model);
```
Comment (arnaud-23) [inline tests/Fixtures/Classes/Api/Controller/PatchFunctionalEntityController.php]: follow this example OC\ApiBundle\Controller\LearningContent\Content\PatchExternalResourceController

this portion of code do the generation of the model and the validation and some other security stuff :
$model = $this->getModelFromRequest(PatchFunctionalEntityModel::class);

File: tests/Fixtures/Classes/Api/Controller/PatchFunctionalEntityController.php [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+     * @param int                   $functionalEntityId
+     * @param PatchFunctionalEntity $model
+     *
+     * @return EditFunctionalEntityRequest
+     */
+    private function buildRequest(
+        int $functionalEntityId,
+        PatchFunctionalEntity $model
+    ): EditFunctionalEntityRequest {
+        $requestBuilder = $this->get(EditFunctionalEntityRequestBuilder::class)
```
Comment (arnaud-23) [inline tests/Fixtures/Classes/Api/Controller/PatchFunctionalEntityController.php]: this class should be passed from the constructor like the builder on the delete controller here
https://github.com/OpenClassrooms/CodeGenerator/pull/29/files#diff-cc28ade8938bd3062b8e8cf550265a79R21

File: tests/Fixtures/Classes/Api/Controller/PostFunctionalEntityController.php [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+    /**
+     * @var CreateFunctionalEntityRequestBuilder
+     */
+    private $createFunctionalEntityRequestBuilder;
+
+    public function __construct(
+        FunctionalEntityViewModelDetailAssembler $assembler,
+        CreateFunctionalEntityRequestBuilder $builder
+    ) {
+        $this->availabilityViewModelDetailAssembler = $assembler;
```
Comment (arnaud-23) [inline tests/Fixtures/Classes/Api/Controller/PostFunctionalEntityController.php]:
```suggestion
        $this->functionalEntityViewModelDetailAssembler = $assembler;
```

File: tests/Fixtures/Classes/Api/Controller/PostFunctionalEntityController.php [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+     */
+    public function postAction(): JsonResponse
+    {
+        try {
+            /** @var PostFunctionalEntity $model */
+            $model = $this->getModelFromRequest(PostFunctionalEntity::class);
+            $response = $this->createFunctionalEntity($model);
+
+            return $this->createCreatedResponse($this->generateLocationUrl(), $this->buildVM($response));
+        } catch (\Exception $e) {
```
Comment (arnaud-23) [inline tests/Fixtures/Classes/Api/Controller/PostFunctionalEntityController.php]: FunctionalEntityNotFoundException

File: tests/Fixtures/Classes/Api/Controller/PostFunctionalEntityController.php [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+            // TODO : set exception
```
Comment (arnaud-23) [inline tests/Fixtures/Classes/Api/Controller/PostFunctionalEntityController.php]: $this->throwNotFoundException();

File: tests/Fixtures/Classes/Api/Controller/PostFunctionalEntityController.php [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+                ->withField2($model->field2)
+                ->withField3($model->field3)
+                ->withField4((new CarbonImmutable(($model->field4))))
+                ->build()
+        );
+    }
+
+    private function generateLocationUrl(): string
+    {
+        return $this->generateUrl('');
```
Comment (arnaud-23) [inline tests/Fixtures/Classes/Api/Controller/PostFunctionalEntityController.php]:
```suggestion
        return $this->generateUrl('', []);
```

File: tests/Fixtures/Classes/Api/Models/Domain/SubDomain/AbstractFunctionalEntity.php [hunk 18]
```diff
@@ -0,0 +1,44 @@
+<?php declare(strict_types=1);
+
+namespace OpenClassrooms\CodeGenerator\Tests\Fixtures\Classes\Api\Models\Domain\SubDomain;
+
+use Symfony\Component\Validator\Constraints as Assert;
+
+abstract class AbstractFunctionalEntity
```
Comment (arnaud-23) [inline tests/Fixtures/Classes/Api/Models/Domain/SubDomain/AbstractFunctionalEntity.php]: no abstract, trait ok but never share code via an abstract it's a mistake

File: tests/Fixtures/Classes/Api/Models/Domain/SubDomain/AbstractFunctionalEntity.php [hunk 19]
```diff
@@ -0,0 +1,44 @@
...
+
+    /**
+     * @var string[]
+     */
+    public $field2;
+
+    /**
+     * @var bool
+     *
+     * @Assert\NotBlank
```
Comment (arnaud-23) [inline tests/Fixtures/Classes/Api/Models/Domain/SubDomain/AbstractFunctionalEntity.php]: for a boolean only:
```suggestion
     * @Assert\NotNull
```

File: tests/Fixtures/Classes/Api/Models/Domain/SubDomain/AbstractFunctionalEntity.php [hunk 20]
```diff
@@ -0,0 +1,44 @@
...
+    protected const FIELD_2_API_FIELD_NAME = 'field2';
+
+    protected const FIELD_3_API_FIELD_NAME = 'field3';
+
+    protected const FIELD_4_API_FIELD_NAME = 'field4';
+
+    /**
+     * @var string
+     *
+     * @Assert\Type("string")
```
Comment (arnaud-23) [inline tests/Fixtures/Classes/Api/Models/Domain/SubDomain/AbstractFunctionalEntity.php]: you can add
@Assert\NotBlank

File: tests/Fixtures/Classes/Api/Models/Domain/SubDomain/PatchFunctionalEntity.php [hunk 21]
```diff
@@ -0,0 +1,49 @@
+<?php declare(strict_types=1);
+
+namespace OpenClassrooms\CodeGenerator\Tests\Fixtures\Classes\Api\Models\Domain\SubDomain;
+
+class PatchFunctionalEntity extends AbstractFunctionalEntity
```
Comment (arnaud-23) [inline tests/Fixtures/Classes/Api/Models/Domain/SubDomain/PatchFunctionalEntity.php]: PatchFunctionalEntityModel

File: tests/Fixtures/Classes/Api/Models/Domain/SubDomain/AbstractFunctionalEntity.php [hunk 22]
```diff
@@ -0,0 +1,44 @@
...
+
+abstract class AbstractFunctionalEntity
+{
+    protected const FIELD_1_API_FIELD_NAME = 'field1';
+
+    protected const FIELD_2_API_FIELD_NAME = 'field2';
+
+    protected const FIELD_3_API_FIELD_NAME = 'field3';
+
+    protected const FIELD_4_API_FIELD_NAME = 'field4';
```
Comment (arnaud-23) [inline tests/Fixtures/Classes/Api/Models/Domain/SubDomain/AbstractFunctionalEntity.php]: move the constants to the Patch

File: tests/Fixtures/Classes/Api/Models/Domain/SubDomain/PostFunctionalEntity.php [hunk 23]
```diff
@@ -0,0 +1,7 @@
+<?php declare(strict_types=1);
+
+namespace OpenClassrooms\CodeGenerator\Tests\Fixtures\Classes\Api\Models\Domain\SubDomain;
+
+class PostFunctionalEntity extends AbstractFunctionalEntity
```
Comment (arnaud-23) [inline tests/Fixtures/Classes/Api/Models/Domain/SubDomain/PostFunctionalEntity.php]: PostFunctionalEntityModel

File: tests/Fixtures/Classes/Api/Models/Domain/SubDomain/AbstractFunctionalEntity.php (hunk 18 above)
Comment (samokiss) [reply tests/Fixtures/Classes/Api/Models/Domain/SubDomain/AbstractFunctionalEntity.php]: nice I didn't know that thank

**Pull request merged**